### PR TITLE
Deployment cleanup

### LINKED ISSUE #1514: SSR API requests failing
Currently, all SSR API requests are failing with an Unknown Error:

HTTP Request
```
{
  request: HttpRequest {
    url: 'https://api.staging.ecosounds.org/projects/filter',
    body: { paging: [Object], filter: {} },
    reportProgress: false,
    withCredentials: true,
    responseType: 'json',
    method: 'POST',
    headers: HttpHeaders {
      normalizedNames: [Map],
      lazyUpdate: null,
      headers: [Map],
      lazyInit: null
    },
    context: HttpContext { map: Map(0) {} },
    params: HttpParams {
      updates: null,
      cloneFrom: null,
      encoder: HttpUrlEncodingCodec {},
      map: Map(0) {}
    },
    urlWithParams: 'https://api.staging.ecosounds.org/projects/filter'
  }
}
```

HTTP Response
```
{
  response: HttpErrorResponse {
    headers: HttpHeaders {
      normalizedNames: Map(0) {},
      lazyUpdate: null,
      headers: Map(0) {}
    },
    status: 0,
    statusText: 'Unknown Error',
    url: 'https://api.staging.ecosounds.org/projects/filter',
    ok: false,
    name: 'HttpErrorResponse',
    message: 'Http failure response for https://api.staging.ecosounds.org/projects/filter: 0 Unknown Error',
    error: ProgressEvent {
      type: 'error',
      target: [XMLHttpRequest],
      currentTarget: [XMLHttpRequest],
      lengthComputable: false,
      loaded: 0,
      total: 0
    }
  }
}
```

My current guess is that this is some form of CORS issue, however without debugging tools (https://github.com/angular/universal/issues/2198) isolating and fixing the bug appears to be fairly difficult

### PR DESCRIPTION
**Pull request by Allcharles**
# Deployment Cleanup

This is a collection of fixes for bugs which were exposed during the staging testing of workbench-client

## Changes

- Cleanup header component unit tests
- Removed pill colouring from (none) values when showing tags associated with an annotation in statistics page
- Hide WIP wrapped components in production build
- Fixed transition issue between SSR and browser where images would still have the loading css class applied after loading
- Fixed issue where SSR was unable to make API requests
- Extracted timeout values for API requests to environment file which can be more easily changed
  - Will need to check if current production timeout values for SSR are long enough to be viable
- Implemented `RehydrationInterceptorService` which will cache the requests made in SSR, and send them to the browser
- Minor changes to express server
  - Using compression middleware to compress data and reduce loading times (TODO validate this improves things)
  - Set X-Frame-Options header to SAMEORIGIN to reduce chances of click-jacking

## Problems

- Need to implement tests for `RehydrationInterceptorService`
- There is a (hopefully) dev only problem where baw-client throws an error attempting to access any of its assets because the local path does not exist

## Issues

Fixes #1514 
#1163 Should now be possible to implement properly

### Statistics Page Before

image

### Statistics Page After

image

## Final Checklist

- [ ] Assign reviewers if you have permission
- [ ] Assign labels if you have permission
- [ ] Link issues related to PR
- [ ] Ensure project linter is not producing any warnings (`npm run lint`)
- [ ] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -44,6 +44,12 @@ LABEL maintainer="Charles Alleman <alleman@qut.edu.au>" \
   ref=${GIT_COMMIT} \
   schema-version="1.0"
 
+USER root
```
Comment (atruskie) [inline Dockerfile]: this should not be necessary - you haven't dropped privileges yet

File: server.ts [hunk 2]
```diff
@@ -50,9 +51,21 @@ export function app(path: string): express.Express {
     })
   );
 
+  // Gzip static assets
+  server.use(compressionModule());
```
Comment (atruskie) [inline server.ts]: We use nginx for this in production. 

We should be ensuring we're using x-sendfile though to serve assets (https://www.nginx.com/resources/wiki/start/topics/examples/xsendfile/) and follow up in our deploy playbooks. For another time though.

File: src/app/helpers/app-initializer/app-initializer.ts [hunk 3]
```diff
@@ -80,7 +80,16 @@ export interface Keys {
  */
 export interface Configuration {
   kind: "Configuration";
+  /** Is the current environment running in production mode */
   production: boolean;
+  /** Timeout for web requests in server side renderer */
+  ssrTimeout: number;
+  /** Timeout for web requests in browser */
+  browserTimeout: number;
```
Comment (atruskie) [inline src/app/helpers/app-initializer/app-initializer.ts:88]: nit: specifically _API_ requests only

**Review by atruskie — CHANGES_REQUESTED** [review]
I think the compression should be removed - nginx should be doing that for us.

I think the rehydration module is the wrong solution to a problem. If the problem being fixed is:

- user sees SSR data
- then loading spinner replaces page
- then data actually loads

Then the solution IMO is to get rid of the obnoxious loading spinner and let the page stream updates as the happen (ideal). Or to rephrase: if data rendered, spinner size := 10% && position := to the side somewhere. (or something similar... our loading bar on the old client was a bar that ran across the top of the page, not to obnoxious). We really only want a large spinner when the area is completely empty.


As pointed out in our discussion: for this to work properly and not force users to some kind of navigate/ refresh to see context-correct data (i.e. API query for logged in state) then the hydrate state feature is broken as is. 
We could ensure express is forwarding the cookie when SSR makes requests, but as we haven't formalised our auth process yet, it seems this is not the right time to make such a decision.

On review: I think save the hydration interceptor for another time and focus on the UI looking nice for now.

File: src/app/helpers/app-initializer/app-initializer.ts (hunk 3 above)
Comment (Allcharles) [reply src/app/helpers/app-initializer/app-initializer.ts:88]: This affects all requests which go through the angular HttpClient library, does not specifically discriminate against API requests

**Review by atruskie — APPROVED** [review]

**Pull request merged**